pkg/gateway/sig/v4_streaming_reader.go

```diff
@@ -184,20 +225,65 @@ func (cr *s3ChunkedReader) Read(buf []byte) (n int, err error) {
 			cr.readS3ChunkHeader()
 			// If we're at the end of a chunk.
 			if cr.n == 0 && cr.err == io.EOF {
-				cr.state = readChunkTrailer
+				cr.state = readChunkTerm
 				cr.lastChunk = true
 				continue
 			}
 			if cr.err != nil {
 				return 0, cr.err
 			}
 			cr.state = readChunk
-		case readChunkTrailer:
-			cr.err = readCRLF(cr.reader)
-			if cr.err != nil {
-				return 0, errMalformedEncoding
+		case readChunkTerm:
+			err = peekCRLF(cr.reader)
+			isTrailingChunk := cr.n == 0 && cr.lastChunk
+
+			if !isTrailingChunk {
+				if readErr := readCRLF(cr.reader); readErr != nil {
+					cr.err = readErr
+					return 0, cr.err
+				}
+				cr.err = err
+			} else if err != nil && !errors.Is(err, errMalformedEncoding) {
+				cr.err = err
+			} else if err == nil {
+				// Some clients send an additional CRLF after the chunk trailer.
+				// This is not part of the spec, but we handle it gracefully.
+				if readErr := readCRLF(cr.reader); readErr != nil {
+					cr.err = readErr
+					return 0, cr.err
+				}
+			}
+
+			// If we're using unsigned streaming upload, there is no signature to verify at each chunk.
+			if cr.chunkSignature != "" {
+				cr.state = verifyChunk
+			} else if cr.lastChunk {
+				cr.state = readTrailerChunk
+			} else {
+				cr.state = readChunkHeader
+			}
+		case readTrailerChunk:
+			checksumAlgorithm, checksumValue := getTrailerChecksum(cr.reader)
+
+			if checksumAlgorithm != cr.checksumAlgorithm {
+				cr.err = ErrChecksumTypeMismatch
+				return 0, cr.err
+			}
+
+			computedChecksum := cr.checksumWriter.Sum(nil)
+			b64checksum := base64.StdEncoding.EncodeToString(computedChecksum)
+			if b64checksum != checksumValue {
+				cr.err = ErrChecksumMismatch
+				return 0, cr.err
+			}
+
+			// Read the final CRLF sequence
+			if err := readCRLF(cr.reader); err != nil {
+				cr.err = err
+				return n, err
```

Comment:
Is there a reason, we are returning a value on error?
```suggestion
				return 0, err
```